k2/csrc/fsa.h

```diff
@@ -122,6 +126,139 @@ struct Fsa {
   }
 };
 
+
+/*
+  Cfsa is a 'const' FSA, which we'll use as the input to operations.  It is
+  designed in such a way that the storage underlying it may either be an Fsa
+  (i.e. with std::vectors) or may be some kind of tensor (probably CfsaVec).
+  Note: the pointers it holds aren't const for now, because there may be
+  situations where it makes sense to change them (even though the number of
+  states and arcs can't be changed).
+ */
+struct Cfsa {
+  int32_t num_states;  // number of states including final state.  States are
+                       // numbered `0 ... num_states - 1`.  Start state is 0,
+                       // final state is state `num_states - 1`.  We store a
+                       // redundant representation here out of a belief that it
+                       // might reduce the number of instructions in code.
+  int32_t begin_arc;   // a copy of arc_indexes[0]; gives the first index in
+                       // `arcs` for the arcs in this FSA.  Will be >= 0.
+  int32_t end_arc;     // a copy of arc_indexes[num_states]; gives the
+                       // one-past-the-last index in `arcs` for the arcs in this
+                       // FSA.  Will be >= begin_arc.
+
+  const int32_t *arc_indexes;   // an array, indexed by state index, giving the
+                                // first arc index of each state.  The last one
+                                // is repeated, so for any valid state 0 <= s <
+                                // num_states we can use arc_indexes[s+1].  That
+                                // is: elements 0 through num_states (inclusive)
+                                // are valid.  CAUTION: arc_indexes[0] may be
+                                // greater than zero.
+
+
+  Arc *arcs;   // Note: arcs[BeginArcIndex()] through arcs[EndArcIndex() - 1]
+               // are valid.
+
+  // Constructor from Fsa
+  Cfsa(const Fsa &fsa);
+
+  Cfsa &operator = (const Cfsa &cfsa) = default;
+  Cfsa(const Cfsa &cfsa) = default;
+
+  Cfsa(int32_t size, int32_t *data);
```

Comment:
Is the format of `data` the same as that of `CfsaVec`?

Reply:
will delete this constructor